pkg/rewrites/rewrites.go

```diff
@@ -0,0 +1,83 @@
+// Package rewrites provides primitives for working with rewrite rules.
+package rewrites
+
+import (
+	"nanotube/pkg/conf"
+	"nanotube/pkg/rec"
+	"regexp"
+
+	"github.com/pkg/errors"
+)
+
+// Rewrites represent all the routing rewrites/routing table.
+type Rewrites []Rewrite
+
+// Rewrite is a routing rewrite.
+type Rewrite struct {
+	From string
+	To   string
+	Copy bool
+
+	CompiledFrom *regexp.Regexp
+}
+
+// Build reads rewrite rules from config, compiles them.
+func Build(crw conf.Rewrites) (Rewrites, error) {
+	var rw Rewrites
+	for _, cr := range crw.Rewrite {
+		r := Rewrite{
+			From: cr.From,
+			To:   cr.To,
+			Copy: cr.Copy,
+		}
+
+		rw = append(rw, r)
+	}
+
+	err := rw.Compile()
+	if err != nil {
+		return rw, errors.Wrap(err, "rewrite rule compilation failed :")
+	}
+
+	return rw, nil
+}
+
+// Compile precompiles regexps
+func (rw Rewrites) Compile() error {
+	for i, r := range rw {
+		cre, err := regexp.Compile(r.From)
+		if err != nil {
+			return errors.Wrapf(err, "compiling rewrite regex: %s failed", cre)
+		}
+		rw[i].CompiledFrom = cre
+	}
+
+	return nil
+}
+
+// RewriteMetric executes all rewrite rules on a record
+// If copy is true and rule matches, we generate new record
+func (rw Rewrites) RewriteMetric(record *rec.Rec) ([]*rec.Rec, error) {
+	result := make([]*rec.Rec, 1)
+	result[0] = record
+
+	for _, r := range rw {
+		newPath := r.CompiledFrom.ReplaceAllString(record.Path, r.To)
+		if newPath == record.Path {
```

Comment:
This may be heavy. Maybe consider matching instead of replace + string check?

Reply:
Decided to do during perf optimization phase.

Reply:
Done.
Performance depends a lot on real data: input/percentage of matches, etc.